File: src/android/DocumentEvidenceCheck.java
package com.nfcdocumentreader;

import android.graphics.Bitmap;
import android.util.Log;

import com.google.android.gms.tasks.Tasks;
import com.google.mlkit.vision.common.InputImage;
import com.google.mlkit.vision.text.Text;
import com.google.mlkit.vision.text.TextRecognizer;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

/**
 * Decides whether a captured photograph actually shows the document, before it is kept.
 *
 * WHY
 * The capture screen photographed whatever was in front of the lens. A real payload from this
 * plugin carried a chip read of a genuine Algerian ID with, attached to it, a photograph of a
 * panda mug and a photograph of a keyboard. The back office received signed chip data and two
 * pictures of an office desk, and nothing in the flow noticed.
 *
 * WHAT DISCRIMINATES, MEASURED ON REAL CAPTURES
 * Two obvious signals turn out not to work. Text volume does not: recognising an ID card yielded
 * 8-9 lines, and the keyboard yielded 8. Face detection does not either: the card's MRZ side has
 * no detectable face at the framing people actually use — a card held in one hand fills perhaps a
 * third of the frame, so its portrait is a hundred pixels across — and a colleague sitting
 * opposite can supply a face that has nothing to do with the document.
 *
 * What does separate them is the document's own machine-readable structure, and the identifiers
 * the flow already knows:
 *
 *   - MRZ lines. Monospaced, 30/36/44 characters of A-Z 0-9 '<'. They survive OCR at the
 *     resolution these photographs are taken at, and nothing on a desk looks like them.
 *   - Identifiers from the MRZ scan and the chip: document number, personal number, surname. On
 *     the real card both sides carried at least one, in print large enough to read.
 *
 * The second is the stronger claim, and it is worth being precise about what it means. It does not
 * establish "this is an ID card". It establishes "this is the document we just read", which is the
 * question actually worth answering: it also rejects a photograph of a different genuine ID.
 *
 * WHAT THIS CANNOT DO
 * It cannot tell a card from a photograph of a card, or from a card displayed on a screen. Those
 * are presentation attacks on the document, and detecting them needs specialist capture — glare
 * and moiré analysis, hologram behaviour under a flash — that this plugin does not have. A
 * "confirmed" result means the right text was in the frame, not that the object was genuine.
 */
final class DocumentEvidenceCheck {

    private static final String TAG = "DocumentEvidence";

    /** Shortest MRZ line is TD1's 30; longest is TD3's 44. Allow for OCR over- and under-run. */
    private static final int MRZ_MIN_LENGTH = 28;
    private static final int MRZ_MAX_LENGTH = 48;
    /** An identifier shorter than this matches too much text by accident. */
    private static final int MIN_IDENTIFIER_LENGTH = 5;

    private DocumentEvidenceCheck() {
    }

    static class Result {
        /** "confirmed" when the document was recognised, "notConfirmed" otherwise. */
        String status = "notConfirmed";
        int textLines = 0;
        int mrzLines = 0;
        String mrzFormat;                       // set when the lines parse as a whole MRZ
        final List<String> matchedIdentifiers = new ArrayList<>();
        final List<String> reasons = new ArrayList<>();

        boolean isConfirmed() {
            return "confirmed".equals(status);
        }

        JSONObject toJson() throws Exception {
            JSONObject json = new JSONObject();
            json.put("status", status);
            json.put("textLines", textLines);
            json.put("mrzLines", mrzLines);
            json.put("mrzFormat", mrzFormat != null ? mrzFormat : JSONObject.NULL);
            // Which identifiers were found, never their values: the values are holder data and
            // already elsewhere in the payload.
            json.put("matchedIdentifiers", new JSONArray(matchedIdentifiers));
            json.put("reasons", new JSONArray(reasons));
            return json;
        }
    }

    /**
     * @param bitmap      the captured frame at full resolution, before compression
     * @param expected    identifiers already known from the MRZ scan or the chip, labelled
     *                    "name:value" so the result can say which one matched without repeating it
     * @param recognizer  ML Kit text recognition
     * @param mrzProcessor used to try to parse the recognised lines as a whole MRZ
     */
    static Result inspect(Bitmap bitmap, List<String> expected,
                          TextRecognizer recognizer, MrzOcrProcessor mrzProcessor) {
        Result result = new Result();
        if (bitmap == null || recognizer == null) {
            result.reasons.add("CHECK_NOT_RUN");
            return result;
        }

        Text text;
        try {
            text = Tasks.await(recognizer.process(InputImage.fromBitmap(bitmap, 0)));
        } catch (Exception e) {
            // A failed check must not read as a failed document.
            Log.w(TAG, "Text recognition failed during the document check: "
                    + e.getClass().getSimpleName());
            result.reasons.add("TEXT_RECOGNITION_FAILED");
            return result;
        }

        List<String> lines = new ArrayList<>();
        for (Text.TextBlock block : text.getTextBlocks()) {
            for (Text.Line line : block.getLines()) {
                lines.add(line.getText());
            }
        }

        String mrzFormat = null;
        try {
            // A whole parsed MRZ is better evidence than loose MRZ-shaped lines, so try for it.
            MrzOcrProcessor.MrzParseResult parsed = mrzProcessor.processText(text);
            if (parsed != null && parsed.isSuccess()) mrzFormat = parsed.format;
        } catch (Exception ignored) {
            // Falls back to the line-shape count in decide().
        }

        return decide(lines, mrzFormat, expected);
    }

    /**
     * The decision itself, separated from how the text was obtained so it can be exercised against
     * real photographs without a device: recognition produces lines, this turns lines into a
     * verdict.
     */
    static Result decide(List<String> lines, String mrzFormat, List<String> expected) {
        Result result = new Result();
        result.mrzFormat = mrzFormat;

        StringBuilder all = new StringBuilder();
        for (String line : lines) {
            result.textLines++;
            if (looksLikeMrzLine(line)) result.mrzLines++;
            all.append(line).append('\n');
        }

        if (result.textLines == 0) {
            result.reasons.add("NO_TEXT_FOUND");
            return result;
        }

        String haystack = normalise(all.toString());
        if (expected != null) {
            for (String labelled : expected) {
                int colon = labelled.indexOf(':');
                if (colon < 1 || colon == labelled.length() - 1) continue;
                String label = labelled.substring(0, colon);
                String needle = normalise(labelled.substring(colon + 1));
                if (needle.length() < MIN_IDENTIFIER_LENGTH) continue;
                if (haystack.contains(needle)) result.matchedIdentifiers.add(label);
            }
        }

        if (!result.matchedIdentifiers.isEmpty() || result.mrzFormat != null || result.mrzLines > 0) {
            result.status = "confirmed";
        } else {
            // Deliberately not graded by how much text there is. A keyboard photographed at a desk
            // produced as many lines as the ID card did, so "plenty of text" is not evidence.
            result.reasons.add(expected == null || expected.isEmpty()
                    ? "NO_MRZ_FOUND"
                    : "NO_MRZ_OR_KNOWN_IDENTIFIER_FOUND");
        }
        return result;
    }

    /** Monospaced, filler-padded, and from the MRZ alphabet — a shape nothing on a desk has. */
    private static boolean looksLikeMrzLine(String line) {
        if (line == null) return false;
        String cleaned = line.replace(" ", "").toUpperCase();
        if (cleaned.length() < MRZ_MIN_LENGTH || cleaned.length() > MRZ_MAX_LENGTH) return false;
        if (cleaned.indexOf('<') < 0) return false;

        int allowed = 0;
        for (char c : cleaned.toCharArray()) {
            if ((c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == '<') allowed++;
        }
        return allowed * 10 >= cleaned.length() * 9;      // at least 90% MRZ alphabet
    }

    /** Case, spacing and punctuation carry no meaning for this comparison. */
    private static String normalise(String value) {
        if (value == null) return "";
        StringBuilder out = new StringBuilder();
        for (char c : value.toUpperCase().toCharArray()) {
            if (Character.isLetterOrDigit(c)) out.append(c);
        }
        return out.toString();
    }
}

File: src/ios/DocumentEvidenceCheck.swift
import Foundation
import UIKit
import Vision

/// Decides whether a captured photograph actually shows the document. Mirrors
/// DocumentEvidenceCheck.java, including what it deliberately does not use.
///
/// The Android note carries the full reasoning; the short version is that the two obvious signals
/// were measured on real captures and do not work. Text volume does not separate a document from a
/// desk — an ID card produced 8-9 recognised lines and a photographed keyboard produced 8. Face
/// detection does not either: the MRZ side of a card has no detectable face at the framing people
/// actually use, while a colleague sitting opposite supplies one that means nothing.
///
/// What separates them is the MRZ's shape, and the identifiers the flow already knows. The second
/// is the stronger claim: it establishes not "this is a document" but "this is the document just
/// read", which also rejects a photograph of a different genuine one.
///
/// It cannot tell a card from a photograph of a card, or from one on a screen — that needs capture
/// this plugin does not have. "confirmed" means the right text was in the frame.
enum DocumentEvidenceCheck {

    private static let mrzMinLength = 28
    private static let mrzMaxLength = 48
    private static let minIdentifierLength = 5

    struct Result {
        var status = "notConfirmed"
        var textLines = 0
        var mrzLines = 0
        var mrzFormat: String?
        var matchedIdentifiers: [String] = []
        var reasons: [String] = []

        var isConfirmed: Bool { return status == "confirmed" }

        var dictionary: [String: Any] {
            return [
                "status": status,
                "textLines": textLines,
                "mrzLines": mrzLines,
                "mrzFormat": mrzFormat ?? NSNull(),
                // Which identifiers matched, never their values: those are holder data and are
                // already elsewhere in the payload.
                "matchedIdentifiers": matchedIdentifiers,
                "reasons": reasons
            ]
        }
    }

    /// Recognises the text, then decides. `expected` holds "label:value" identifiers.
    static func inspect(_ image: UIImage, expected: [String]) -> Result {
        guard let cgImage = image.cgImage else {
            var result = Result()
            result.reasons.append("CHECK_NOT_RUN")
            return result
        }

        let request = VNRecognizeTextRequest()
        request.recognitionLevel = .accurate
        request.usesLanguageCorrection = false

        do {
            try VNImageRequestHandler(cgImage: cgImage, options: [:]).perform([request])
        } catch {
            // A failed check must not read as a failed document.
            var result = Result()
            result.reasons.append("TEXT_RECOGNITION_FAILED")
            return result
        }

        let lines = (request.results ?? []).compactMap {
            ($0 as? VNRecognizedTextObservation)?.topCandidates(1).first?.string
        }
        return decide(lines: lines, mrzFormat: MrzOcrProcessor().processLines(lines)?.format,
                      expected: expected)
    }

    /// The decision itself, separated from how the text was obtained so it can be exercised
    /// against real photographs without a device.
    static func decide(lines: [String], mrzFormat: String?, expected: [String]) -> Result {
        var result = Result()
        result.mrzFormat = mrzFormat
        result.textLines = lines.count

        guard !lines.isEmpty else {
            result.reasons.append("NO_TEXT_FOUND")
            return result
        }

        result.mrzLines = lines.filter { looksLikeMrzLine($0) }.count
        let haystack = normalise(lines.joined(separator: "\n"))

        for labelled in expected {
            guard let colon = labelled.firstIndex(of: ":") else { continue }
            let label = String(labelled[labelled.startIndex..<colon])
            let needle = normalise(String(labelled[labelled.index(after: colon)...]))
            guard label.count > 0, needle.count >= minIdentifierLength else { continue }
            if haystack.contains(needle) { result.matchedIdentifiers.append(label) }
        }

        if !result.matchedIdentifiers.isEmpty || result.mrzFormat != nil || result.mrzLines > 0 {
            result.status = "confirmed"
        } else {
            // Deliberately not graded by how much text there is: a keyboard photographed at a desk
            // produced as many lines as the ID card did, so "plenty of text" is not evidence.
            result.reasons.append(expected.isEmpty
                ? "NO_MRZ_FOUND"
                : "NO_MRZ_OR_KNOWN_IDENTIFIER_FOUND")
        }
        return result
    }

    /// Monospaced, filler-padded, from the MRZ alphabet — a shape nothing on a desk has.
    private static func looksLikeMrzLine(_ line: String) -> Bool {
        let cleaned = line.replacingOccurrences(of: " ", with: "").uppercased()
        guard cleaned.count >= mrzMinLength, cleaned.count <= mrzMaxLength,
              cleaned.contains("<") else { return false }

        let allowed = cleaned.unicodeScalars.filter {
            ($0 >= "A" && $0 <= "Z") || ($0 >= "0" && $0 <= "9") || $0 == "<"
        }.count
        return allowed * 10 >= cleaned.count * 9        // at least 90% MRZ alphabet
    }

    /// Case, spacing and punctuation carry no meaning for this comparison.
    private static func normalise(_ value: String) -> String {
        return value.uppercased().unicodeScalars
            .filter { CharacterSet.alphanumerics.contains($0) }
            .map(String.init).joined()
    }
}
